src/devtools/views/Components/TreeContext.js

```diff
@@ -596,6 +596,20 @@ function TreeContextController({ children, viewElementSource }: Props) {
           state = reduceTreeState(store, state, action);
           state = reduceSearchState(store, state, action);
           state = reduceOwnersState(store, state, action);
+
+          // If the selected ID is in a collapsed subtree, reset the selected index to null.
+          // We'll know the correct index after the layout effect will toggle the tree,
+          // and the store tree is mutated to account for that.
+          if (
+            state.selectedElementID !== null &&
+            store.isInsideCollapsedSubTree(state.selectedElementID)
+          ) {
```

Comment:
This makes me nervous.

Reply:
Not sure what you mean specifically

Reply:
This change makes me nervous- temporarily setting an incorrect state and relying on an effect's cascading update to fix it up for us. Also because of the potential for impact with the two `setState` stuff I'm working on over in PR #196.

Reply:
Hmm. I wouldn't say it's "incorrect".

I'd say the state is "incorrect" when index refers to a different item than is selected. Which is what used to happen.

In this case we're explicitly acknowledging that `selectIndex` is *unknown*. It's only incorrect if other code can't deal with that — but afaik it does.

Reply:
Here's another concrete example where modeling unknown value explicitly makes a difference.

Before this fix, if you selected a node, collapsed its parent, and pressed "down", the selection would "jump" way below: 

![Screen Recording 2019-04-22 at 10 26 PM](https://user-images.githubusercontent.com/810438/56532523-d1e76200-654d-11e9-80b0-9f2d580cd045.gif)

This is because it would still be using the index — and the index is not valid anymore! So we end up in a difference place in the tree.

With this fix, pressing "down" when selected node is collapsed always navigates you to the root element.

I'm not saying this is necessarily a better behavior (maybe it should navigate to the closest non-collapsed parent, or maybe it should expand instead). But I'm saying that this fix exposes an existing problem which doesn't go away even if we didn't merge it.

Reply:
Hm. I think that sounds like a case where our `HANDLE_STORE_MUTATION` action just does the wrong thing. Rather than checking if the item is still in the tree, it should check to see if it's still in the *visible* tree, and if not, bump selection up to its nearest visible ancestor.

Reply:
That would be great!

Reply:
However that also reveals a broader problem. If you a node whose child is selected gets deleted, we currently completely lose the selection. However instead we should select the closest non-deleted (and non-collapsed while we’re at it) parent. Or sibling. But definitely not lose the selection like we do now.

Reply:
> If a node whose child is selected gets deleted, we currently completely lose the selection.

I don't think I understand. The child would also be _explicitly_ deleted in such a case, right?

Reply:
Sorry, I misspoke. I mean this:

Select node
Delete that node

Ideal behavior: selection either goes to sibling or to parent.

Current behavior: selection disappears.

I wanted to fix this by selecting parent. But by the time node is removed we can’t query store about its parent anymore. I guess we could fix this by making removedIds a map that suggests next selection to the tree context mutation handler.

Reply:
When you say  "current behavior"  do you mean prior to this change landing? or current-current?